### ISSUE TITLE
tryUpdateNodeHealth may process old data in large-scale cluster scenarios.

### ISSUE THREAD

**Issue by Black-max12138** [kubernetes/kubernetes#128643]
### What happened?

https://github.com/kubernetes/kubernetes/blob/154b756e2ed850d2e64baea269dbb749ac02a77d/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L711-L729
The node information used by tryUpdateNodeHealth is obtained from nodes, err := nc.nodeLister.List(labels.Everything()). If there are a large number of nodes and the processing capability of kube-controller-manager is limited, tryUpdateNodeHealth may take a long time to process. The node information that causes the follow-up processing is not the latest, so I think this is a problem.

### What did you expect to happen?

tryUpdateNodeHealth should be the latest data each time it processes it.

### How can we reproduce it (as minimally and precisely as possible)?

Large-scale nodes

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```
1.31
</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by aojea** [comment]
Yeah, this was parallelized here https://github.com/kubernetes/kubernetes/pull/114296

The only option I can think about is to expose the number of workers so people can adjust the concurrecny for their needs

https://github.com/kubernetes/kubernetes/blob/c9024e7ae628f1473a6cac28e7bd6cd8e64f872f/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L293-L295

/sig scalability

**Comment by Black-max12138** [comment]
> Yeah, this was parallelized here #114296
> 
> The only option I can think about is to expose the number of workers so people can adjust the concurrecny for their needs
> 
> https://github.com/kubernetes/kubernetes/blob/c9024e7ae628f1473a6cac28e7bd6cd8e64f872f/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L293-L295
> 
> /sig scalability

Check whether the latest node can be obtained before tryUpdateNodeHealth.

**Comment by googs1025** [comment]
I'm curious if there is any real scenario? How big a cluster might this problem occur? 🤔

**Comment by AnishShah** [comment]
sig-node CI meeting notes:

Even though this is a bug, tuning `nodeUpdateWorkerSize` needs some exploration as it increases resource usage.

/triage accepted
/remove-kind bug
/kind feature
